[Autograd Onboarding Lab] Adding custom autograd function for 'attn' and verified with gradcheck

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62125
* __->__ #62124

Part 2/4 of onboarding lab. This PR will not be merged to master.

Part 1 is [here](https://colab.research.google.com/drive/1kV6waOHR-j6WyNv_OvxaqCLMAbvThOQr#scrollTo=oVkOL4gUm-Am).

NivekT committed:
Adding custom autograd function for 'attn' and verified with gradcheck

[ghstack-poisoned]

albanD commented on test/test_attn.py:
```diff
@@ -0,0 +1,68 @@
+        tanh_output = torch.tanh(mm_output)
+        output = torch.mm(tanh_output, v)
+
+        ctx.save_for_backward(q, k, v, mm_output, tanh_output)
```
As mentioned offline, using mm_output here means that `gradgradcheck` will fail. You want to fix that I think.

NivekT replied:
Fixed.

albanD commented on test/test_attn.py:
```diff
@@ -0,0 +1,68 @@
+        grad_q = grad_k = grad_v = None
+
+        if ctx.needs_input_grad[0]:  # q
+            dtanh_out = 1. - (torch.tanh(mm_output) ** 2)
```
It would be nice to re-use computations between the different cases here.
In particular, depending on which ones you decide to implement together for your c++ implementation, you can do the same collapsing here and test all the re-use in advance (before writing them in c++).

NivekT replied:
Good feedback, I changed the code accordingly.

albanD reviewed (commented):
Could you add that file to `run_test.py` so that it actually runs in CI and we can see in the PR if it all passes properly?

NivekT committed:
Update on "[Autograd Onboarding Lab] Adding custom autograd function for 'attn' and verified with gradcheck"


Part 2/4 of onboarding lab. This PR will not be merged to master.

Part 1 is [here](https://colab.research.google.com/drive/1kV6waOHR-j6WyNv_OvxaqCLMAbvThOQr#scrollTo=oVkOL4gUm-Am).

[ghstack-poisoned]

NivekT commented:
Closing this because this was done (the remaining CI failures are because of merging with master branch, confirmed with @albanD).
